src/firefly/test/edu/caltech/ipac/visualize/plot/HistogramTest.java

```diff
@@ -0,0 +1,362 @@
+        int goodpix = 0;
+        for (int i = 0; i < HISTSIZ2; i++)
+            goodpix += histArray[i];
+
+        int goal = (int) (goodpix * (ra_value) / 100);
+        int sum=0;
+        int i=-1;
+        do {
+            i++;
+            sum+=histArray[i];
+        } while(sum<goal);
+        if (round_up)
+            return ((i + 1.0) * histBinsize + histMin);
+        else
+            return ((i) * histBinsize + histMin);
+    }
+
+    /**
+     * Build the hist array to be used for calculating expected values
+     * @param data
+     * @param histMin
+     * @param histBinsize
+     * @return
+     */
+    private static int[] buildHistArray(float[] data, double histMin, double histBinsize){
+
+        int [] histArray = new int[HISTSIZ2 + 1];
+        for (int k = 0; k < data.length; k++) {
+
+            if (!Double.isNaN(data[k])) {
+
+                int i = (int) ((data[k] - histMin) / histBinsize);
+                if (i < 0 || i > HISTSIZ2) {
+                    continue;
+                } else {
+                    histArray[i]++;
+                }
+
+            }
+        }
+        return histArray;
+
+    }
+    private double[] jsonArray1dToDouble1d(JSONArray jsonArray1d){
+        double[] double1d = new double[jsonArray1d.size()];
+        for (int i=0; i<jsonArray1d.size(); i++){
+            double1d[i]=(double)jsonArray1d.get(i);
+        }
+        return double1d;
+    }
+
+    /**
+     * This method validate the end - to - end test results.
+     * @param hist
+     * @param jsonObject
+     */
+    private void validate(Histogram hist, JSONObject jsonObject ){
+        double rangeValue = FitsRead.getDefaultRangeValues().getUpperValue();
+        Assert.assertEquals(hist.get_pct(rangeValue, true), jsonObject.get("pct"));
+        Assert.assertEquals(hist.get_sigma(rangeValue, true), jsonObject.get("sigma"));
+
+        JSONArray jArray=  (JSONArray) jsonObject.get("tbl");
+        double[] expectedTbl = jsonArray1dToDouble1d(jArray);
+        Assert.assertArrayEquals(hist.getTblArray(),expectedTbl, delta );
+        System.out.println("Histogram end to end test is pass") ;
+
+    }
+    @Test
+    /**
+     * This method is an end to end test.  It takes a FITS file, f3.fits and creates an instance of Histogram.
+     * Compare its results with the ones stored in the f3HistogramResult.json file.  The results stored in
+     * f3HistogramResult.json was generated by the same procedures at the time when HistogramTest class is added.
+     * It ensures that any future changes will not introduce problems.
+     */
+    public void testEndToEnd() throws FitsException, IOException, ParseException {
+        String testTreePath = HistogramTest.class.getProtectionDomain().getCodeSource().getLocation().getPath();
+        String  rootPath = testTreePath.split("firefly")[0];
+        String dataPath = rootPath+TEST_DATA_PATH;
+        String inFitsName =dataPath+fileName;
```

Comment:
If there is a common method of resolving these paths, will you create a TestUtil class to collect them?  I see them repeated over and over.